libraries/bot-applicationinsights/src/test/java/com/microsoft/bot/applicationinsights/TelemetryWaterfallTests.java

```diff
@@ -0,0 +1,396 @@
+        .assertReply("step2")
+        .send("hello")
+        .assertReply("step3")
+        .startTest()
+        .join();
+
+        // C#'s trackEvent method of BotTelemetryClient has nullable parameters,
+        // therefore it always calls the same method.
+        // On the other hand, Java's BotTelemetryClient overloads the trackEvent method,
+        // so instead of calling the same method, it calls a method with less parameters.
+        // In this particular test, WaterfallDialog's beginDialog calls the method with only two parameters
+        verify(telemetryClient, times(4)).trackEvent(
+            anyString(),
+            anyMap()
+        );
+    }
+
+    @Test
+    public void waterfallWithCallback() {
+        ConversationState convoState = new ConversationState(new MemoryStorage());
+        TestAdapter adapter = new TestAdapter(TestAdapter.createConversationReference("waterfallWithCallback", "User1", "Bot"))
+            .use(new AutoSaveStateMiddleware(convoState));
+
+        BotTelemetryClient telemetryClient = Mockito.mock(BotTelemetryClient.class);
+        StatePropertyAccessor<DialogState> dialogState = convoState.createProperty("dialogState");
+        DialogSet dialogs = new DialogSet(dialogState);
+        WaterfallDialog waterfallDialog = new WaterfallDialog("test", newWaterfall());
+
+        dialogs.add(waterfallDialog);
+        dialogs.setTelemetryClient(telemetryClient);
+
+        new TestFlow(adapter, turnContext -> {
+            DialogContext dc = dialogs.createContext(turnContext).join();
+            dc.continueDialog().join();
+            if (!turnContext.getResponded()) {
+                dc.beginDialog("test", null).join();
+            }
+            return CompletableFuture.completedFuture(null);
+        })
+        .send("hello")
+        .assertReply("step1")
+        .send("hello")
+        .assertReply("step2")
+        .send("hello")
+        .assertReply("step3")
+        .startTest()
+        .join();
+
+        // C#'s trackEvent method of BotTelemetryClient has nullable parameters,
+        // therefore it always calls the same method.
+        // On the other hand, Java's BotTelemetryClient overloads the trackEvent method,
+        // so instead of calling the same method, it calls a method with less parameters.
+        // In this particular test, WaterfallDialog's beginDialog calls the method with only two parameters
+        verify(telemetryClient, times(4)).trackEvent(
+            anyString(),
+            anyMap()
+        );
+    }
+
+    @Test
+    public void waterfallWithActionsNull() {
+        BotTelemetryClient telemetryClient = Mockito.mock(BotTelemetryClient.class);
+        WaterfallDialog waterfall = new WaterfallDialog("test", null);
+        waterfall.setTelemetryClient(telemetryClient);
+        try {
+            waterfall.addStep(null);
+            Assert.fail();
+        } catch (IllegalArgumentException ex) {
+            ex.printStackTrace();
+        }
```

Comment:
You are not validating the exception correctly, check this [example](https://github.com/microsoft/botbuilder-java/blob/4dd06e088f45616ef5bd1aae140e6726bd682332/libraries/bot-dialogs/src/test/java/com/microsoft/bot/dialogs/prompts/ActivityPromptTests.java#L37)